onnxruntime/core/optimizer/insert_cast_transformer.cc

```diff
@@ -37,30 +42,115 @@ onnxruntime::NodeArg* AddCastNode(onnxruntime::Graph& graph,
   return new_arg;
 }
 
-static bool IsInputFloat16(const onnxruntime::Node& node) {
-  for (auto input : node.InputDefs()) {
-    if (input->Type() != nullptr &&
-        DataTypeImpl::TypeFromProto(*input->TypeAsProto()) == DataTypeImpl::GetTensorType<MLFloat16>() &&
-        !node.GetExecutionProviderType().empty()) {
-      return true;
-    }
+// check if the node has an fp16 input but was not able to be assigned an execution provider.
+// we will need to add a casts to/from fp32 around the node for it to be executed
+static bool NodeNeedsInputCastToFp32(const onnxruntime::Node& node) {
+  bool not_assigned = node.GetExecutionProviderType().empty();
+
+  if (not_assigned) {
+    const auto& input_defs = node.InputDefs();
+    bool has_fp16_input = std::any_of(input_defs.cbegin(), input_defs.cend(),
+                                      [](const NodeArg* input_def) {
+                                        return IsMLFloat16Tensor(*input_def);
+                                      });
+    return has_fp16_input;
   }
+
   return false;
 }
 
-static bool IsSingleInputNodeFloat16Node(const onnxruntime::Node& node) {
-  if (IsInputFloat16(node) && node.GetExecutionProviderType() == kCpuExecutionProvider) {
-    for (auto it = node.InputNodesBegin(); it != node.InputNodesEnd(); ++it) {
-      if (IsInputFloat16(*it))
-        return false;
-    }
-    for (auto it = node.OutputNodesBegin(); it != node.OutputNodesEnd(); ++it) {
-      if (IsInputFloat16(*it))
-        return false;
-    }
-    return true;
+// Detect an isolated node that is able to process fp16 data but is between other nodes that have fp16 inputs
+// but will need a Cast inserted to enable them to run.
+//
+// Say we have 3 nodes in the middle of a graph that all have fp16 inputs.
+//
+// -> NodeA -> NodeB -> NodeC ->
+//
+// NodeA and NodeC have no kernel that can handle fp16 data (no execution provider assigned).
+//   e.g. 'Add' does not have an fp16 kernel
+// NodeB has a kernel that can process fp16 data (assigned to CPU EP).
+//
+// By default, we would insert Cast to/from fp32 around NodeA and NodeC as all operators have an fp32 kernel.
+//
+// i.e. -> CastToFp32 -> NodeA -> CastToFp16 -> NodeB -> CastToFp32 -> NodeC -> CastToFp16
+//
+// We can avoid the casts around NodeB if we also force that to run using fp32 data.
+//
+// Detect this scenario by checking the input and output edges of the node for fp16 values to that are coming from or
+// going to a node that will need a Cast.
+//
+// Return true if all the fp16 inputs and outputs are connected to nodes that will be cast to fp32.
+static bool IsIsolatedFp16Node(const onnxruntime::Node& node, onnxruntime::Graph& graph) {
+  bool isolated_fp16_node = false;
+
+  // if node has input coming from other nodes (only consuming graph inputs or initializers if it doesn't),
+  //    does not have a subgraph (would have to alter subgraph inputs if we cast the input to this node),
```

Comment:
nit: formatting ?

Reply:
was deliberate indenting - one line per condition of the 'if' code that it is documenting